drivers/watchdog/virtio_wdt.c

```diff
@@ -0,0 +1,240 @@
+	struct virtio_device *vdev;
+	struct virtqueue *vq;
+	struct completion have_data;
+	char name[25];
+	unsigned int data_avail;
+	int index;
+	bool busy;
+	bool wdd_register_done;
+};
+
+static void virtio_watchdog_recv_done(struct virtqueue *vq)
+{
+	struct virtio_watchdog_info *vi = vq->vdev->priv;
+
+	/* We can get spurious callbacks, e.g. shared IRQs + virtio_pci. */
+	if (!virtqueue_get_buf(vi->vq, &vi->data_avail))
+		return;
+
+	complete(&vi->have_data);
+}
+
+/* Host will change the buffer from 0->1 */
+static void register_buffer(struct virtio_watchdog_info *vi, u8 *buf,
+							size_t size)
+{
+	struct scatterlist sg;
+
+	sg_init_one(&sg, buf, size);
+	virtqueue_add_inbuf(vi->vq, &sg, 1, buf, GFP_KERNEL);
+
+	virtqueue_kick(vi->vq);
+}
+
+
+static int virtio_watchdog_ping(struct watchdog_device *wdd)
+{
+	struct virtio_watchdog_info *vi = to_info(wdd);
+	struct virtio_device *vdev = vi->vdev;
+	int ret;
+	u8 *buf;
+
+	if (!vi->wdd_register_done)
+		return -ENODEV;
+
+	buf = kzalloc(sizeof(u8), GFP_KERNEL);
+	if (!vi->busy) {
+		vi->busy = true;
+		reinit_completion(&vi->have_data);
+		register_buffer(vi, (void *)buf, sizeof(u64));
+	}
+
+	ret = wait_for_completion_killable(&vi->have_data);
+	if (ret < 0)
+		goto err;
+
+	if (*buf != 1) {
+		dev_err(&vdev->dev,
+				"Host did not acknowledge buffer correctly");
+		ret = -EINVAL;
+	}
+
+err:
+	vi->busy = false;
+	kfree(buf);
+
+	return ret;
+}
+
+static int virtio_watchdog_start(struct watchdog_device *wdd)
+{
+	struct virtio_watchdog_info *vi = to_info(wdd);
+	struct virtio_device *vdev = vi->vdev;
+
+	dev_info(&vdev->dev, "Watchdog started");
+
+	return 0;
+}
+
+static int virtio_watchdog_stop(struct watchdog_device *wdd)
+{
+	struct virtio_watchdog_info *vi = to_info(wdd);
+	struct virtio_device *vdev = vi->vdev;
+
+	dev_info(&vdev->dev, "Watchdog stop request ignored");
+
+	return 0;
+}
+
+static struct watchdog_info vw_info = {
+	.identity = "virtio-watchdog",
+	.options = WDIOF_KEEPALIVEPING,
+};
+
+static const struct watchdog_ops vw_ops = {
+	.owner = THIS_MODULE,
+	.start = virtio_watchdog_start,
+	.stop = virtio_watchdog_stop,
+	.ping = virtio_watchdog_ping,
+};
+
+static int probe_common(struct virtio_device *vdev)
+{
+	int err, index;
+	struct virtio_watchdog_info *vi = NULL;
+
+	vi = kzalloc(sizeof(struct virtio_watchdog_info), GFP_KERNEL);
+	if (!vi)
+		return -ENOMEM;
+
+	vi->index = index =
+		ida_simple_get(&watchdog_index_ida, 0, 0, GFP_KERNEL);
+	if (index < 0) {
+		err = index;
+		goto err_ida;
+	}
+	sprintf(vi->name, "virtio_watchdog.%d", index);
+	init_completion(&vi->have_data);
+
+	vdev->priv = vi;
+	vi->vdev = vdev;
+
+	vi->vq =
+		virtio_find_single_vq(vdev, virtio_watchdog_recv_done, "input");
+	if (IS_ERR(vi->vq)) {
+		err = PTR_ERR(vi->vq);
+		goto err_find;
+	}
+
+	vi->wdd.info = &vw_info;
+	vi->wdd.ops = &vw_ops;
+	vi->wdd.min_timeout = VW_HEARTBEAT_DEFAULT;
+	vi->wdd.max_timeout = VW_HEARTBEAT_DEFAULT;
+	vi->wdd.timeout = VW_HEARTBEAT_DEFAULT;
```

Comment:
From reading `drivers/watchdog/watchdog_dev.c`, it seems the timeout can still be changed by the user if `watchdog_set_timeout()` is being called, right? (This is fine, I just want to make sure I understand the code correctly)

Reply:
You need to advertise a capability to indicate if that ioctl is supported.